profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LinuxStackFramesCollector.cpp

```diff
@@ -177,83 +216,26 @@ void LinuxStackFramesCollector::NotifyStackWalkCompleted(std::int32_t resultErro
     _stackWalkFinished = true;
 }
 
-void LinuxStackFramesCollector::InitializeSignalHandler()
-{
-    if (s_isSignalHandlerSetup)
-        return;
-
-    std::unique_lock<std::mutex> lock{s_signalHandlerInitLock};
-
-    if (s_isSignalHandlerSetup)
-        return;
-
-    s_isSignalHandlerSetup = SetupSignalHandler();
-}
-
-bool LinuxStackFramesCollector::TrySetHandlerForSignal(int32_t signal, struct sigaction& action)
-{
-    struct sigaction oldAction;
-    if (sigaction(signal, nullptr, &oldAction) < 0)
-    {
-        Log::Error("LinuxStackFramesCollector::TrySetHandlerForSignal:"
-                   " Unable to examine signal handler for signal ",
-                   signal, ". Reason:",
-                   strerror(errno), ".");
-        return false;
-    }
-
-    // Replace signal only if there is no user-defined one.
-    // @ToDo: is that enough?
-    if (oldAction.sa_handler == SIG_DFL || oldAction.sa_handler == SIG_IGN)
-    {
-        sigaddset(&action.sa_mask, signal);
-        int32_t result = sigaction(signal, &action, &oldAction);
-        if (result == 0)
-        {
-            return true;
-        }
-
-        sigdelset(&action.sa_mask, signal);
-        Log::Error("LinuxStackFramesCollector::TrySetHandlerForSignal:"
-                   " Unable to setup signal handler for signal",
-                   signal, ". Reason: ",
-                   strerror(errno), ".");
-    }
-
-    Log::Info("LinuxStackFramesCollector::TrySetHandlerForSignal:"
-              " Unable to set signal for ",
-              signal, ". The default one is overriden by ",
-              oldAction.sa_handler, ".");
-
-    return false;
-}
-
 bool LinuxStackFramesCollector::SetupSignalHandler()
 {
-    // SIGUSR1 & SIGUSR2 are not use in the CLR
-    // But, let's check if they are available
-
     struct sigaction sampleAction;
-    sampleAction.sa_flags = 0;
-    sampleAction.sa_handler = LinuxStackFramesCollector::CollectStackSampleSignalHandler;
+    sampleAction.sa_flags = SA_RESTART | SA_SIGINFO;
+    sampleAction.sa_handler = SIG_DFL;
+    sampleAction.sa_sigaction = LinuxStackFramesCollector::CollectStackSampleSignalHandler;
     sigemptyset(&sampleAction.sa_mask);
+    sigaddset(&sampleAction.sa_mask, s_signalToSend);
 
-    if (TrySetHandlerForSignal(SIGUSR1, sampleAction))
-    {
-        s_signalToSend = SIGUSR1;
-        Log::Info("LinuxStackFramesCollector::SetupSignalHandler: Successfully setup signal handler for SIGUSR1 signal.");
-        return true;
-    }
-
-    if (TrySetHandlerForSignal(SIGUSR2, sampleAction))
+    int32_t result = sigaction(s_signalToSend, &sampleAction, &s_previousAction);
+    if (result != 0)
     {
-        s_signalToSend = SIGUSR2;
-        Log::Info("LinuxStackFramesCollector::SetupSignalHandler: Successfully setup signal handler for SIGUSR2 signal.");
-        return true;
+        sigdelset(&sampleAction.sa_mask, s_signalToSend);
```

Comment:
What's the point of calling `sigdelset` here? 🤔

Reply:
artifact from the past :P